fetcher/fetcher.go

```diff
@@ -0,0 +1,228 @@
+			if errors.Is(err, database.ErrNotFound) {
+				f.update(t.key, nil, false, 0)
+				continue
+			} else if err != nil {
+				f.stopOnce.Do(func() {
+					f.err = err
+					close(f.stop)
+				})
+				return
+			}
+
+			numChunks, ok := keys.NumChunks(v)
+			if !ok {
+				f.stopOnce.Do(func() {
+					f.err = ErrInvalidKeyValue
+					close(f.stop)
+				})
+				return
+			}
+			f.update(t.key, v, true, numChunks)
+		case <-f.stop:
+			return
+		}
+	}
+}
+
+// Checks if a key is in the cache. If it's not in cache, check
+// if we were the first to request the key.
+func (f *Fetcher) shouldFetch(t *task) bool {
+	f.keyLock.RLock()
+	defer f.keyLock.RUnlock()
+
+	if f.keysToFetch[t.key].cache != nil {
+		return false
+	}
+	return f.keysToFetch[t.key].queue[0] == t.id
+}
+
+// Updates the cache and dependencies
+func (f *Fetcher) update(k string, v []byte, exists bool, chunks uint16) {
+	f.keyLock.Lock()
+	// Puts a key that was fetched from disk into cache
+	f.keysToFetch[k].cache = &fetchData{v, exists, chunks}
+	queue := f.keysToFetch[k].queue
+	f.keysToFetch[k].queue = nil
+	f.keyLock.Unlock()
+
+	f.txnLock.Lock()
+	for _, id := range queue {
+		f.txnsToFetch[id].Done() // Notify all other txs
+	}
+	f.txnLock.Unlock()
+}
+
+// Lookup enqueues keys for the workers to fetch
+// Invariant: Don't call [Lookup] afer calling [Stop]
+func (f *Fetcher) Lookup(ctx context.Context, txID ids.ID, stateKeys state.Keys) *sync.WaitGroup {
+	wg := &sync.WaitGroup{}
+
+	f.keyLock.Lock()
+	tasks := make([]*task, 0, len(stateKeys))
+	for k := range stateKeys {
+		if _, ok := f.keysToFetch[k]; !ok {
+			f.keysToFetch[k] = &keyData{cache: nil, queue: make([]ids.ID, 0)}
+		}
+		// Don't requeue keys that were already fetched
+		if f.keysToFetch[k].cache != nil {
+			continue
+		}
+		f.keysToFetch[k].queue = append(f.keysToFetch[k].queue, txID)
+		t := &task{
+			ctx: ctx,
+			id:  txID,
+			key: k,
+		}
+		tasks = append(tasks, t)
+	}
+	wg.Add(len(tasks))
+	f.keyLock.Unlock()
+
+	f.txnLock.Lock()
+	f.txnsToFetch[txID] = wg
+	f.txnLock.Unlock()
+
+	for _, t := range tasks {
+		f.fetchable <- t
+	}
+	return wg
+}
+
+func (f *Fetcher) Get(wg *sync.WaitGroup, stateKeys state.Keys) (map[string]uint16, map[string][]byte) {
+	// Block until all keys for the tx are fetched
+	wg.Wait()
+
+	f.l.Lock()
+	f.completed++
+	f.l.Unlock()
+
+	f.keyLock.Lock()
+	defer f.keyLock.Unlock()
+
+	// Fetch keys from cache
+	var (
+		reads   = make(map[string]uint16, len(stateKeys))
+		storage = make(map[string][]byte, len(stateKeys))
+	)
+	for k := range stateKeys {
+		if v := f.keysToFetch[k].cache; v != nil {
+			reads[k] = v.chunks
+			if v.exists {
+				storage[k] = v.v
+			}
+		}
+	}
+	return reads, storage
+}
+
+func (f *Fetcher) Stop() {
+	f.stopOnce.Do(func() {
+		f.err = ErrStopped
+		close(f.stop)
+	})
+}
+
+// Wait until all the workers are done and return any errors
+func (f *Fetcher) Wait() error {
+	f.l.Lock()
+	if f.completed == f.numTxs {
```

Comment:
This isn't a stable cleanup. You should close `f.fetchable` once all items are enqueued, not by this condition (as most of the time you won't close it then). You don't need to track this.